keps/sig-apps/3322-add-maxRestartTimes-to-podSpec/README.md

```diff
@@ -0,0 +1,946 @@
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This KEP aims to add a new field named `maxRestartTimesOnFailure` to podSpec which helps to control
+the restart times of a Pod when Pod's `restartPolicy` is set to `OnFailure`.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Currently, we have `backoffLimit` to fail a Job after a certain amount of retries, which helps to
+prevent unnecessary restarts. `Pod` is also eager for this feature to avoid infinite restarting
+due to a system internal error etc. What's more, considering `Job` or other workloads are working
+at a higher level, adding `maxRestartTimesOnFailure` to `Pod` might be reusable.
+
+As a matter of fact, it's a long-standing issue and strongly needed by the community, see [issue](https://github.com/kubernetes/kubernetes/issues/65797).
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- Add a new field `maxRestartTimesOnFailure` to `pod.spec` which only applies when `restartPolicy` is set to `OnFailure`
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Combine the implementations of `backoffLimit` and `maxRestartTimes` together.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Add a new field `maxRestartTimesOnFailure` to the `pod.spec`. `maxRestartTimesOnFailure` only applies
+when `restartPolicy` is set to `OnFailure`. As a result, the restart times of a Pod
+will be controlled by `maxRestartTimesOnFailure`.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+When running with a pod, some system errors or program internal errors
+encountered, which causes the application restarting infinitely. This is
+unrecoverable and wasteful.
+
+### Story 2
+
+For batch Jobs specifically, now a Job provides a single backoffLimit for all derived pods,
```

Comment:
cc @mimowo

Reply:
I see. In that KEP I propose to scope the backoff-limit per index to when restartPolicy=Never, because in this case Job controller handles the restarts. I guess the paragraph (in the current form) fits better into the Notes section.

Reply:
I'll add supporting `restartPolicy=Never` as a discussion point.